R-package/R/lgb.Predictor.R

```diff
@@ -126,6 +127,110 @@ Predictor <- R6::R6Class(
         num_row <- nrow(preds)
         preds <- as.vector(t(preds))
 
+      } else if (predcontrib && inherits(data, c("dsparseMatrix", "dsparseVector"))) {
+
+        ncols <- .Call(LGBM_BoosterGetNumFeature_R, private$handle)
+        ncols_out <- integer(1L)
+        .Call(LGBM_BoosterGetNumClasses_R, private$handle, ncols_out)
+        ncols_out <- (ncols + 1L) * max(ncols_out, 1L)
+        if (!inherits(data, "dsparseVector") && ncols_out > .Machine$integer.max) {
+          stop("Resulting matrix of feature contributions is too large for R to handle.")
+        }
+
+        if (inherits(data, "dsparseVector")) {
+
+          if (length(data) > ncols) {
```

Comment:
For every `stop()` that you've included in this PR, can you please add a unit test checking that that specific error is raised, using `testhat::expect_error()`?

Here's an example:

https://github.com/microsoft/LightGBM/blob/3798f8720f4fbc4f1249717fb8f2f037e0075a46/R-package/tests/testthat/test_basic.R#L523-L537

I've found tests like that are especially useful for methods like this one with many `if - else if` branches, to be sure that refactorings don't result in unreachable code paths or accidentally failing to raise an informative error.

Reply:
I don't think that's a desirable behavior. The package doesn't support other types of sparse inputs *right now*, but if in the future it were to add support for e.g. COO matrices, one would not want to find out that there are tests out there specifically testing that it fails for COO matrices.

Testing all of these conditions would also imply fitting the model to a CSC matrix with close to 2^30 columns.

Reply:
> Testing all of these conditions would also imply fitting the model to a CSC matrix with close to 2^30 columns

You're right, saying "every `stop()`" was too much. It's ok to not test the conditions like `.Machine$integer.max`. 

But please add these types of tests as requested for each of the "shape of input data" `stop()`  calls (`if (ncol(data) > ncols) `) and for the one on supported types like `stop(sprintf("Predictions on sparse inputs are only allowed for...`.

Reply:
In that case it'd make more sense to have general tests about unexpected number of columns in the data for all prediction types. The current code for other prediction inputs and types also throws such errors, and those errors don't have tests.

Reply:
I also don't think it'd be appropriate to have tests checking that it fails for some class like symmetric unit-diagonal sparse CSR matrices just for this - it'd make more to have those tests for all prediction possibilities.

Reply:
> The current code for other prediction inputs and types also throws such errors, and those errors don't have tests.

I am trying to establish the practice that *changes* to this project are accompanied by tests on those *changes*, so that at least the amount of untested code in the project isn't increased by new PRs. You're not wrong about there being other code paths in the project that are untested or under-tested, but that fact isn't a reason to avoid adding tests in this or other new PRs.

> to have those tests for all prediction possibilities

I don't understand this second point, sorry. I'm not suggesting that LightGBM's tests need an exhaustive list of possible unsupported data types that might be passed into it.

Reply:
You are asking to add tests for `predict(<sparse matrix>, predcontrib=TRUE)`. The conditions to test are not specific to `<sparse matrix>`, nor to `predcontrib=TRUE`. If tests like those were to be added, the logical structure would be to add them for different possible call types to `predict(<data>, <types>)`, not just for `predict(<sparse matrix>, predcontrib=TRUE)`.

Reply:
I'm limiting what I'm asking for to only `predict(<sparse matrix>, predcontrib=TRUE)` because those are the only cases being added in this PR and as I said, I'm trying to avoid any *new* untested code being added to the project.

If you want to create a separate PR adding tests on these types of shape checks for all combinations of `<data>` and `<type>` you're welcome to and it would be helpful, but I'm not asking you to commit that time and effort for the purpose of this PR.

Reply:
Added the requested tests, but I'd suggest a big refactor after these PRs are merged.